core/src/replay_stage.rs

```diff
@@ -3172,6 +3172,19 @@ impl ReplayStage {
                                         .get_hash(last_voted_slot)
                                         .expect("Must exist for us to have frozen descendant"),
                                 );
+                                // The lockout check is cached for previously computed banks. Since we are
+                                // updating our tower we need to update this cache as well.
+                                for slot in frozen_banks.iter().map(|b| b.slot()) {
+                                    let stats = progress
+                                        .get_fork_stats_mut(slot)
+                                        .expect("All frozen banks must exist in the Progress map");
+                                    stats.is_locked_out = tower.is_locked_out(
+                                        slot,
+                                        ancestors
+                                        .get(&slot)
+                                        .expect("Ancestors map should contain slot for is_locked_out() check"),
+                                    );
```

Comment:
this is a good catch!

Don't all of these tower based checks https://github.com/solana-labs/solana/blob/9558abdcf131579b9eddbdb710c313aba7b16fc8/core/src/replay_stage.rs#L3252-L3261 need to be redone? 

Might be easier to move all those tower-based checks into a separate function that we can call from both places so that we don't miss this in the future for any added checks

Reply:
good catch, added all of these as well